added simple tutorial to introduce the jupyterGIS GUI. (#393)
geojupyter/jupytergis: docs/tutorials/intro.md

Reviewer: #411

# Getting Started with JupyterGIS

Welcome to the first tutorial in the JupyterGIS series! This guide will introduce you to the basics of JupyterGIS and help you set up your environment, create your first map, and understand its key features.

> ## Objectives

Reviewer: :star_struck: https://diataxis.fr/tutorials/#show-the-learner-where-they-ll-be-going

![JupyterGIS User Interface explained](../assets/images/jupyterGIS_interface.png)

The elements shown in the figure above are:

1. The application Launcher helps you to select which application you want to start e.g. a Notebook, a Console, and another application such as a Terminal or open a GIS file (which can be either a JupyterGIS file or a QGIS file).
2. GIS Layers List / Browser Panel
3. Left Sidebar which contains a file browser, a list of tabs in the main work and of running kernels and terminals, the command palette, the table of contents, the extension manager, and the JupyterGIS extension which allows you to see the GIS layers list.
4. The GIS object properties, annotations and filters of a selected GIS layer.
5. The Jupyter toolbar menu which you will use with Jupyter Notebooks.
6. The Log console (which you can you for debugging).
7. Right Sidebar showing the property inspector (active in notebooks), the debugger, and GIS object properties, annotation and filters.

## Adding your first layers

We will start the JupyterGIS application and create a basic map.

**The goal for this lesson: To get started with an example map.**

> :::{admonition} Get JupyterGIS

Reviewer 2: Since you're using colon fences here and we're using myst-parser, you'll have to add the `colon_fence` extension to our `conf.py`. See https://github.com/jupyterlite/xeus/blob/main/docs/conf.py#L9 for an example of this.

Author: Thank you!

   ![Add a Vector Layer](../assets/images/addVectorLayer.png)
4. Set the path of the new shapefile Layer to `https://www2.census.gov/geo/tiger/GENZ2015/shp/cb_2015_us_county_500k.zip` and click **Ok**.
   ![Set Shapefile Layer](../assets/images/CreateShapefileLayer.png)
5. You should then be able to see the new Shapefile Layer on the map and when clicking on the **Custom Shapefile Layer** (see figure below), you should also see the details of the object properties on the right of the map. The map being interactive, you can zoom in and out as you wish to focus on your area of interest (here over the US).
   ![View Shapefile Layer](../assets/images/viewShapefileLayer.png)

:::{admonition} Why don't we add a local shapefile?
:class: tip
One of the main strength of JupyterGIS is its collaborative capabilities. To be able to collaborate, it is much easier to work with data shared and available online.
:::

```{exercise-start}
:label: add-vector-layer
```

- Add a new Shapefile Layer to your map:

```
https://docs.geoserver.org/stable/en/user/_downloads/30e405b790e068c43354367cb08e71bc/nyc_roads.zip

> ```

Reviewer: Both layers display as "Custom Shapefile layer" at this point in the tutorial, and same for the sources. It would be nice to give them names early in this tutorial to reduce the cognitive load of the learner! Then we can demonstrate renaming later in the tutorial without having to pay the cognitive load price.

- Zoom over New-York (USA) and check if you can see the newly added layer.
- Let's customize this new layer by changing the color. In the GIS Layer/Browser Panel, select the top layer (corresponds to the last layer you added) and right click to **Edit Symbology**. Then change the **Stroke Color** to a color of your choice. You can also change the Stroke Width and check the result after pressing \*_OK_.
- In a similar way, you can edit the symbology of the first Shapefile Layer we added and change the **Fill Color**, **Stroke Color** and **Stroke Width**.
- Do you still see the roads in New-York? Try to adjust the **Opacity** value (default is 1) to a lower value for this Shapefile Layer. Can you see all your layers now?

```{exercise-end}

```

> ```{solution} add-vector-layer

Reviewer 3: That being said, I'm totally fine if it's not supported and please do not consider as a blocker or something for getting this in.

> ```{solution} add-vector-layer

Author: Good point. I added the dropdown in the markdown but forgot to add sphinx-togglebutton in the requirements. Thanks!

:class: dropdown

After adding the new Shapefile Layer and zooming over New-York, you should have the following map:
![View several shapefile Layer on the same map](../assets/images/severalShapefileLayers.png)

When you right click to edit the symbology, you should get the following pop-up menu:
![Edit symbology](../assets/images/editSymbology.png)

And when editing the symbology of the other Shapefile Layer:
![Adjust Symbology](../assets/images/adjustSymbology.png)

> And updating the opacity (for instance to 0.5) of the US county Shapefile Layer, you can get a map that can look like to (depending on the colors and Stroke width you chose!):

Reviewer: ```suggestion
And updating the opacity (for instance to 0.5) of the US county Shapefile Layer, you can get a map that looks like this (depending on the colors and Stroke width you chose!):
```

![final map for the intro exercise](../assets/images/finalIntroMap.png)
```

## Hiding and Reordering Layers

### Hiding/Viewing Layers

In the GIS Layers List/Browser Panel, you can select a layer and click the **eye** icon to hide or show the corresponding layer.
![Hide a layer](../assets/images/hideLayer.png)

### Reordering Layers

The layers in your Layers List are displayed on the map according to their order in the list. The bottom layer is drawn first, while the top layer is drawn last. You can adjust their drawing order by rearranging the layers in the list.

The order in which the layers have been loaded into the map is probably not logical at this stage. It’s possible that the road layer or/and the country layer are completely hidden because other layers are on top of it.

For example, this layer order:

> ![Wrong layer order](../assets/images/wrongLayerOrder.png)

Reviewer: I think we should only show layers in this screenshot, not sources. It's a bit confusing!

> would hide all the custom shapefile layers.

Reviewer: ```suggestion
would hide all the custom shapefile layers underneath the OpenStreetMap raster layer.
```

To solve this problem, you can click and drag on a layer in the Layers list to move it up or down, and reordering the layers as you wish them to be viewed on the map.

## Renaming Layers

As shown on the previous image, the two custom shapefile layers we added have the same name in the GIS Layers List / Browser Panel which can be very confusing.
It is a good practice to rename your layers with meaningful names. To rename a layer, select it in the GIS Layers List / Browser Panel, and right click to **Rename Layer**.

```{exercise-start}
:label: Rename-layer
```

> - Rename each **Custom Shapefile Layer** with a meaningful name e.g., **cb_2015_us_county_500k** for the US County Shapefile Layer and **nyc_roads** for the roads in New-York.

Reviewer: Layer names can have spaces in them, so I feel it would be valuable to write these as more human-readable names, like "NYC roads" and "US counties" instead of naming them after the input files. It might be useful to demonstrate renaming the _sources_ to match the input file names!

```{exercise-end}

```

```{solution} Rename-layer
:class: dropdown
![Map with renamed layers](../assets/images/renameLayer.png)
```

## Save your JGIS project

> Your jGIS file corresponds to your GIS project (similar to `.qgz´ for QGIS) and you probably already noticed that the default name is `untitled.jGIS`. If you creater several jGIS files without renaming them, a number will be added before the extension e.g. `untitled1.jGIS`, `untitled2.jGIS`, etc.

Reviewer 2: ```suggestion
Your jGIS file corresponds to your GIS project (similar to `.qgz` for QGIS) and you probably already noticed that the default name is `untitled.jGIS`. If you creater several jGIS files without renaming them, a number will be added before the extension e.g. `untitled1.jGIS`, `untitled2.jGIS`, etc.
```

It is important to give meaningful name to your jGIS project.

To rename it, you can click on **File** in the Jupyter toolbar menu and select **Rename JGIS...** to give a new name to your jGIS project.

![Rename jGIS project](../assets/images/renamejGIS.png)

## Conclusion

> Now you’ve learned how to add all the layers you need from various sources and have created a basic map!

Reviewer: We could go into a bit more depth here (repeating some content from the top of the tutorial), and possibly provide some more positivity! What about something along the lines of:

```suggestion
Congratulations on completing this tutorial! You've created a basic map in JupyterGIS, learned how to add new layers, rearrange layers, alter symbology, and save your project. You're now well on your way to becoming a pro JupyterGIS user!
```

**The objective of this lesson is to understand the fundamentals of the JupyterGIS user interface.**
:::

![JupyterGIS User Interface explained](../assets/images/jupyterGIS_interface.png)

The elements shown in the figure above are:

1. The application Launcher helps you to select which application you want to start e.g. a Notebook, a Console, and another application such as a Terminal or open a GIS file (which can be either a JupyterGIS file or a QGIS file).
2. Left Sidebar which contains a file browser, a list of tabs in the main work and of running kernels and terminals, the command palette, the table of contents, the extension manager, and the JupyterGIS extension which allows you to see the GIS layers list.
3. GIS Layers List / Browser Panel
4. Right Sidebar showing the property inspector (active in notebooks), the debugger, and GIS object properties, annotation and filters.
5. The GIS object properties, annotations and filters of a selected GIS layer.
6. The Jupyter toolbar menu which you will use with Jupyter Notebooks.
7. The Log console (which you can you for debugging).

## Adding your first layers

We will start the JupyterGIS application and create a basic map.

> **The goal for this lesson: To get started with an example map.**

Reviewer: ```suggestion
:::{admonition} Lesson goal
:class: tip
**The goal for this lesson: To get started with an example map.**
:::
```

:::{admonition} Get JupyterGIS
:class: tip
Before beginning this exercise, JupyterGIS must be installed on your computer (see [Installation instructions](https://geojupyter.github.io/jupytergis/install.html)) or you can use an online version of JupyterGIS (such as [![Jupyterlite badge](https://jupyterlite.rtfd.io/en/latest/_static/badge.svg)](https://geojupyter.github.io/jupytergis/lite/lab/index.html?path=france_hiking.jGIS/)).

> :::

Reviewer: Nitpick: Maybe this should be at the top of the document? :thinking:

Let's rename the source called 'Custom Shapefile Source' to 'US Counties' so that we can easily identify it in the list when adding a new source.
![Rename 'Custom Shapefile Source' to 'US Counties'](../assets/images/renameSource.png)

```{exercise}
:label: Rename-layer

- Rename the **Custom Shapefile Layer** with a meaningful name e.g., **US Counties** for the US County Shapefile Layer.

```

```{solution} Rename-layer
:class: dropdown
![Rename 'Custom Shapefile Layer' to 'US Counties'](../assets/images/renameLayer.png)
```

:::{admonition} Ordering in the list of Sources versus Layers
:class: warning
Did you notice in the figures above that in the list of sources, the source for 'US Counties' appears before the source for 'OpenStreetMap.Mapmik'? In contrast, you can see the reverse order in the list of Layers.
The reason is that sources are ordered alphabetically, while layers are ordered based on the order in which they are added, with the top layer being the most recently added to your map.

> :::

Reviewer: Nice callout!!!
